File: 04_12_Paths_with_Sum.cpp
class Solution {

public:
    void pathSum(TreeNode* root, bool addToSum, long int targetSum, int & currentCount) {
        if (root == nullptr) {
            return;
        }

        if (root->val == targetSum) {
            currentCount++;
        }

        if (addToSum == false) {
            pathSum(root->left, false, targetSum, currentCount);
            pathSum(root->right, false, targetSum, currentCount);
        }

        pathSum(root->left, true, targetSum - root->val, currentCount);
        pathSum(root->right, true, targetSum - root->val, currentCount);
    }

    int pathSum(TreeNode* root, int targetSum) {
        int totalSum = 0;
        this->pathSum(root, false, targetSum, totalSum);
        return totalSum;
    }
};

File: 04_12_Paths_with_Sum.java
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        Map<Integer, Integer> prefixSum = new HashMap<>();
        
        prefixSum.put(0, 1);
        
        return pathSumHelper(root, prefixSum, 0, targetSum);
    }
    
    int pathSumHelper(TreeNode current, Map<Integer, Integer> prefixSum, int currentSum, int targetSum) {
        if (current == null) {
            return 0;
        }
        
        currentSum += current.val;
        
        
        
        int targetSumCount = prefixSum.getOrDefault(currentSum - targetSum, 0);
        
        prefixSum.put(currentSum, prefixSum.getOrDefault(currentSum, 0) + 1);
        
        
        targetSumCount += pathSumHelper(current.left, prefixSum, currentSum, targetSum);
        targetSumCount += pathSumHelper(current.right, prefixSum, currentSum, targetSum);
        
        prefixSum.put(currentSum, prefixSum.get(currentSum) - 1);
        
        return targetSumCount;
        
    }
}
